targets/csharp/source/source/PlayFabGameserverInternal.cs

```diff
@@ -0,0 +1,324 @@
+
+            _heartbeatRunning = true;
+            _signalHeartbeatEvent.Reset();
+            TransitionToActiveEvent.Reset();
+
+            _heartbeatTask = Task.Run(HeartbeatAsync);
+            return Task.CompletedTask;
+        }
+
+        private GameserverConfiguration GetConfiguration()
+        {
+            string fileName;
+
+            if (!string.IsNullOrWhiteSpace(_overrideConfigFileName))
+            {
+                fileName = _overrideConfigFileName;
+            }
+            else
+            {
+                fileName = Environment.GetEnvironmentVariable(PlayFabGameserverSDK.GsdkConfigFileEnvVarKey);
+            }
+
+            GameserverConfiguration localConfig;
+
+            if (!string.IsNullOrWhiteSpace(fileName) && File.Exists(fileName))
+            {
+                var jsonWrapper = PlayFab.PluginManager.GetPlugin<ISerializerPlugin>(PluginContract.PlayFab_Serializer);
+
+                localConfig = new JsonFileConfiguration(jsonWrapper, fileName);
+            }
+            else
+            {
+                localConfig = new EnvironmentVariableConfiguration();
+            }
+
+            return localConfig;
+        }
+
+        private IDictionary<string, string> CreateConfigMap(GameserverConfiguration localConfig)
+        {
+            var finalConfig = new Dictionary<string, string>();
+
+            foreach (KeyValuePair<string, string> certEntry in localConfig.GameCertificates)
+            {
+                finalConfig[certEntry.Key] = certEntry.Value;
+            }
+
+            foreach (KeyValuePair<string, string> metadata in localConfig.BuildMetadata)
+            {
+                finalConfig[metadata.Key] = metadata.Value;
+            }
+
+            foreach (KeyValuePair<string, string> port in localConfig.GamePorts)
+            {
+                finalConfig[port.Key] = port.Value;
+            }
+
+            finalConfig[PlayFabGameserverSDK.HeartbeatEndpointKey] = localConfig.HeartbeatEndpoint;
+            finalConfig[PlayFabGameserverSDK.ServerIdKey] = localConfig.ServerId;
+            finalConfig[PlayFabGameserverSDK.LogFolderKey] = localConfig.LogFolder;
+            finalConfig[PlayFabGameserverSDK.SharedContentFolderKey] = localConfig.SharedContentFolder;
+            finalConfig[PlayFabGameserverSDK.CertificateFolderKey] = localConfig.CertificateFolder;
+            finalConfig[PlayFabGameserverSDK.TitleIdKey] = localConfig.TitleId;
+            finalConfig[PlayFabGameserverSDK.BuildIdKey] = localConfig.BuildId;
+            finalConfig[PlayFabGameserverSDK.RegionKey] = localConfig.Region;
+
+            return finalConfig;
+        }
+
+        private async Task HeartbeatAsync()
+        {
+            while (_heartbeatRunning)
+            {
+                if (_signalHeartbeatEvent.WaitOne(1000))
+                {
+                    if (_debug)
+                    {
+                        Logger.Log("State transition signaled an early heartbeat.");
```

Comment:
Can we make this more descriptive? I remember seeing in the old GSDK but it is not clear what does this indicate

Reply:
how about: Game state transition occurred, new game state is {State}. Sending heartbeat sooner than the configured 1 second interval.

---
In reply to: [255766944](https://github.com/PlayFab/SDKGenerator/pull/398#discussion_r255766944) [](ancestors = 255766944)